README.md
# Table of Contents
1. [Introduction](#intro)
2. [Initial Setup](#setup)
3. [Adapter Requirements & Guidelines](#requirements)
4. [Adapter Package](#adapterPackage)
    * [Adapter Terminology](#adapterTerminology)
    * [Adapter Files](#adapterFiles)
5. [New Adapter](#newAdapter)
    * [Module Setup](#moduleSetup)
    * [Getting Started](#gettingStarted)
    * [Adapter Module Overview](#overview)
        * [Configuration](#configuration)
        * [Event Model](#eventModel)
        * [Creating an Adapter Module](#creatingAdapterModule)
7. [Utility Libraries](#helpers)
    * [Utils](#utils)
    * [BidRoundingTransformer](#bidRounding)
8. [Linting](#linting)
9. [Debugging](#debugging)
10. [Testing](#testing)
11. [Code Submission Guidelines](#codeSubmissionGuidelines)


# <a name='intro'></a>Introduction

<b>Welcome to the IX Library Adapters repository!</b>

This repository contains all the certified IX Library adapters, as well as the tools to develop and verify your own adapter.
Below you will find the information you will need to complete the certification process and become a part of the IX Library Ecosystem!

There are two branches in this repository:
* `master` - The stable branch where all certified adapters reside.
* `canary` - The branch used for adapters to test their updates or new adapter with publishers before merging to the stable `master` branch.

## New Adapters

New adapters must submit a pull request to the `canary` branch upon meeting all the requirements highlighted in the [Requirements](#requirements) section to test their adapter.  After testing is complete, submit a pull request to the `master` branch. The adapter update will then be merged into the stable `master` branch, making it availabe to all publishers using the IX Library.

## Existing Adapters

When updates are made to an existing adapter, a pull request will need to be made to the `canary` branch, so that any updates can be validated in production in a controlled manner. The adapter update will then be merged into the stable master branch, making it availabe to all publishers using the IX Library.

# <a name='setup'></a>Initial Setup
To obtain all the tools you will need to develop your adapter, follow these steps:
1. Fork and clone the `ix-library-adapters` repository
2. In the root directory of `ix-library-adapters` repository, run `npm install` 

This will install all the dependencies and tools provided by Index Exchange to help with the adapter development process.

# <a name='requirements'></a> Adapter Requirements & Guidelines
In order for your module to be successfully certified, please refer to the following list of requirements and guidelines. Items under required <b><u>must</u></b> be satisfied in order to pass the certification process. Items under guidelines are recommended for an optimal setup and should be followed if possible.

### General

##### Required
* The only targeting keys that can be set are predetermined by Index. The adapter module should not be setting targeting on other keys.
* Must support the following browsers: IE 9+, Edge, Chrome, Safari, and Firefox
* Must run system tests in all supported browsers
* Must not use polyfills

##### Guidelines
* Please use our helper libraries when possible. Refer to our [Utility Libraries](#helpers) documentation below. All of the utility functions in this library are designed to be backwards compatible with supported browsers to ease cross-browser compatibility.

### Bid Request Endpoint

##### Required
* Must provide cache busting parameter. Cache busting is required to prevent network caches.
* Adapter endpoint domain must be consistent. Load balancing should be performed by the endpoint.
* Your endpoint should support HTTPS request. When an IX Library loads on secure pages, all requests need to be HTTPS. If you're unable to provide a secure endpoint, the IX Library will not be able to make requests to your ad servers.

##### Guidelines
* Your module should support a single request architecture which has a capability to combine multiple bids into a single HTTP request.
* Adapter should use AJAX to make bid requests. Please use our [Network](#network) utility library for making network requests, particularly the `Network.ajax` method. This ensures the requests will work with all supported browsers.

### Bid Response

##### Required
* Returned bid must be a <b>net</b> bid.
* Bid passes must be explicit in the response.
* All returned bids shall be in the currency trafficked by the publisher.
* Size must be returned in the response. The sizes must also match the requested size. This size parameter will be validated.
* Encrypted prices must be flagged in the response.

### Pixels & Tracking Beacons

##### Required
* Pixels/beacons must only be dropped when the adapter wins the auction and their ad renders.
* Dropping pixels during auction slows down the execution of auctions and is not allowed by Index Exchange.

### DFP Setup

##### Required
* DFP line items, creatives, and render functions will be set up by the Index Exchange team.
<br><br>

# <a name='adapterPackage'></a>Adapter Package

### <a name='adapterTerminology'></a> Adapter Terminology
Below is a list of the terms we will be using throughout this document to explain the process of building your own adapter:

* `<adapter-name>` - This identifies the adapter package for which a given file belongs to. This is used for the package name itself as well as the files within the package e.g. `"example-company"`
* `<partner-name>` - Each adapter is assigned a `partnerId` which is used by the IX Library framework to internally interact with the adapter. It is defined in the adapter module file in the following format: `ExampleCompanyHtb`


### <a name='adapterFiles'></a> Adapter Files
* `<adapter-name>-htb.js` - This is your adapter module file, by default it contains a template divided into multiple sections which need to be completed.
* `<adapter-name>-htb-validator.js` - This is the validator file for the configuration object that will be passed into your module.
* `<adapter-name>-htb-exports.js` - A file used to ensure that we are able to work with legacy creatives and bid responses which require callback functionality.
* `<adapter-name>-htb-system-tests.js` - This file contains functions that will be called when the system tests are run.

# <a name='newAdapter'></a> New Adapter
<b>Note: If you have an existing adapter that has been migrated to the new repository, skip this step</b><br>

### <a name='moduleSetup'></a> Module Setup

Once you have cloned the repository and obtained all the tools as specified above, you can start setting up your new adapter:

* Depending on which OS you're using, you can create a new base adapter template by running one of the following commands:
    * If you're running OSX or Linux, run `npm run new-adapter "Company Name"`
    * If you're running Windows, run `npm run win-new-adapter "Company Name"`

| Name              | Type      | Required  | Description                                                                                                   | Constraints               |
|----------------   |--------   |---------- |-------------------------------------------------------------------------------------------------------------- |-------------------------  |
| "Company Name"| string    | Yes       | The company name including spaces with quotations around it.  | Alphanumeric characters   |

<br>
* For instance, an adapter for "Example Company" would run the following command on a Linux system:
    * `npm run new-adapter "Example Company"`

## <a name='gettingStarted'></a> Getting Started
1. <b>Complete the `<adapter-name>-htb.js` file</b>
    * `<adapter-name>-htb.js` is where all of your adapter code will live.
    * In order to complete the adapter module correctly, please refer to the [Adapter Module Overview](#overview) and the [Utility Libraries](#helpers) sections.
    * <b>Please refer to the [Adapter Requirements and Guidelines](#requirements) when creating your module. Ensure requirements are met to streamline the review process.</b>
2. <b>Complete the `<adapter-name>-htb-validator.js` file</b>
    * This file is where your adapter-specific configurations will need to be validated.
3. <b>Complete the `<adapter-name>-htb-exports.js` file</b>
    * This file is used to support older adapters with legacy creatives and adapters that execute a callback as a response.
    * If your response executes a callback, you must add the following code to your file:<br>

```javascript
if (__directInterface.Layers.PartnersLayer.Partners.ExampleCompanyHtb) {
        shellInterface.ExampleCompanyHtb = shellInterface.ExampleCompanyHtb || {};
        shellInterface.ExampleCompanyHtb.adResponseCallbacks = __directInterface.Layers.PartnersLayer.Partners.ExampleCompanyHtb.adResponseCallbacks;
}
```
4. <b>Complete the `<adapter-name>-htb-system-tests.js` file</b>
    * This file contains all your adapter's test functions. Follow the process outlined on <a href="https://kb.indexexchange.com/ix_library_partners/build_adapter/write_and_run_the_test_cases.htm">Knowledge Base</a> to write your test cases.
5. <b>Run the system tests and linter</b>
    * Make sure all your test cases are passing by running `npm run debug` on OSX or Linux, or `npm run win-debug` on Windows in your adapter folder and then accessing the system tests page.
    * Ensure that there are no lint errors in your adapter folder by running `npm run lint` on OSX or Linux, or `npm run win-lint` on Windows.
6. <b>Verify adapter through the debugger</b>
    * Test your adapter is working as expected by using the debugger as a test page. This can be done by running `npm run debug` on OSX or Linux, or `npm run win-debug` and accessing the debug page.

## <a name='overview'></a> Adapter Module Overview

### <a name='configuration'></a>Prelude: Configuration & Parcels

The IX Library passes each adapter a configuration object that has been configured based on a publisher's website. This object contains all the configuration information required for the IX Library, including adapter-specific slot mappings, timeouts, and any other adapter-specific configuration.
The adapter-specific slot mappings dictate how ad slots on the page will map to adapter-specific configuration.
There are 2 concepts to be familiar with when understanding how slots on the page are mapped back to adapter-specific configuration. Header Tag Slots, which refers to htSlots and adapter-specific configuration, which refers to xSlots in the codebase.
* htSlots - This is an abstraction of the googletag.slot object.
  * These will need to be mapped to xSlots.
* xSlots - These are also an abstraction for adapter-specific configuration that will be mapped to htSlots.
  * These represent a single adapter-specific configuration.
  * An xSlot is how a adapter can map their ad server specific identifiers (placementIDs, siteIDs, zoneIDs, etc) to the `htSlot` object.
  * It can represent a single or multiple sizes.
  * Multiple xSlots can be mapped to the same htSlot.

Example Adapter Configuration Mapping
```javascript
{
    "partners": {
        "<partner-name>": {
            "enabled": true,
            "configs": {
                "xSlots": {
                    "xSlot1": {
                        "placementID": "123",
                        "sizes": [ [300, 250], [300, 600] ]
                    },
                    "xSlot2": {
                        "placementID": "345",
                        "sizes": [ [300, 250] ]
                    }
                },
                "mapping": {
                    "htSlotID-1": [ "xSlot1" ],
                    "htSlotID-2": [ "xSlot2" ]
                }
            }
        }
    }
}
```

Based on the mapping defined in the adapter config, <i>Parcels</i> will be generated for every xSlot/htSlot pair. Parcels are objects that carry different kinds of information throughout the IX Library. In the context of the adapter, parcels are the input into your adapter. Parcels carry information regarding which slots on the publisher's page need demand. More specifically parcels contain information like xSlot reference, htSlot references, demand (after its been applied by the adapter in parseResponse), etc. Each parcel represents a single combination of an htSlot and an xSlot.

Each parcel is an object in the following form:

```javascript
{
    "partnerId": "<partner-name>",
    "htSlot": {
      "__type__": "HeaderTagSlot"
    },
    "ref": "googletag.Slot",      // reference to the slot on the page, in this example it is a googletag slot
    "xSlotRef": {   // contains the reference to the xSlot object from the partner configuration.
      "placementID": "123",
      "sizes": [ [300, 250], [300, 600] ]
    },
    "xSlotName": "1",
    "requestId": "_fwyvecpA" // generated by the IX Library and is used to map the creative back to a piece of demand.
                            // This will show up as targeting under the id key, and is used by the render function.
}
```

These parcels will be fed into both of the functions that your adapter needs to implement:

1. First `generateRequestObj` will need to craft a bid request for the `parcels` that need demand.

2. Second `parseResponse` will take the same parcels and apply the demand (set specific properties in the parcel objects) that is returned from the bid requests to the same parcels and send them back to the IX Library.

## <a name='eventModel'></a> High Level Event Model

1. The IX Library is loaded on the page.
2. IX Library specific configuration validation is performed.
3. All the adapter modules are instantiated.
    * Adapter-specific configuration validation is performed - checking that all the required fields are provided and that they are in the correct format.
4. An external request for demand is made to the IX Library. This can be via a googletag display or refresh call, or by other methods depending on the IX Library in use.
The IX Library requests demand from the adapter modules for the required slots (provided in the form of parcels).
    * The IX Library calls `generateRequestObj(returnParcels)` for every adapter module.
    * The adapter then crafts and returns a request object based on the parcels (containing slot information) specified.
    * The IX Library then sends out a bid request using the request object.

teads/CHANGES.md
# 2.0.1

- Add TCF v2 support (forward apiVersion in bid request)
- Support deal targeting
- Support time to first byte


colossus/colossus-htb.js
'use strict';

////////////////////////////////////////////////////////////////////////////////
// Dependencies ////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////

var Browser = require('browser.js');
var Classify = require('classify.js');
var Constants = require('constants.js');
var Partner = require('partner.js');
var Size = require('size.js');
var SpaceCamp = require('space-camp.js');
var System = require('system.js');
var Network = require('network.js');

var ComplianceService;
var RenderService;

//? if (DEBUG) {
var ConfigValidators = require('config-validators.js');
var PartnerSpecificValidator = require('colossus-htb-validator.js');
var Scribe = require('scribe.js');
var Whoopsie = require('whoopsie.js');
//? }

////////////////////////////////////////////////////////////////////////////////
// Main ////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////

/**
 * Partner module template
 *
 * @class
 */
function ColossusHtb(configs) {
    /* =====================================
     * Data
     * ---------------------------------- */

    /* Private
     * ---------------------------------- */

    /**
     * Reference to the partner base class.
     *
     * @private {object}
     */
    var __baseClass;

    /**
     * Profile for this partner.
     *
     * @private {object}
     */
    var __profile;

    /* =====================================
     * Functions
     * ---------------------------------- */

    /* Utilities
     * ---------------------------------- */

    /**
     * Generates the request URL and query data to the endpoint for the xSlots
     * in the given returnParcels.
     *
     * @param  {object[]} returnParcels
     *
     * @return {object}
     */
    function __generateRequestObj(returnParcels) {
        /* =============================================================================
         * STEP 2  | Generate Request URL
         * -----------------------------------------------------------------------------
         *
         * Generate the URL to request demand from the partner endpoint using the provided
         * returnParcels. The returnParcels is an array of objects each object containing
         * an .xSlotRef which is a reference to the xSlot object from the partner configuration.
         * Use this to retrieve the placements/xSlots you need to request for.
         *
         * If your partner is MRA, returnParcels will be an array of length one. If your
         * partner is SRA, it will contain any number of entities. In any event, the full
         * contents of the array should be able to fit into a single request and the
         * return value of this function should similarly represent a single request to the
         * endpoint.
         *
         * Return an object containing:
         * queryUrl: the url for the request
         * data: the query object containing a map of the query string paramaters
         *
         * callbackId:
         *
         * arbitrary id to match the request with the response in the callback function. If
         * your endpoint supports passing in an arbitrary ID and returning it as part of the response
         * please use the callbackType: Partner.CallbackTypes.ID and fill out the adResponseCallback.
         * Also please provide this adResponseCallback to your bid request here so that the JSONP
         * response calls it once it has completed.
         *
         * If your endpoint does not support passing in an ID, simply use
         * Partner.CallbackTypes.CALLBACK_NAME and the wrapper will take care of handling request
         * matching by generating unique callbacks for each request using the callbackId.
         *
         * If your endpoint is ajax only, please set the appropriate values in your profile for this,
         * i.e. Partner.CallbackTypes.NONE and Partner.Requesttypes.AJAX. You also do not need to provide
         * a callbackId in this case because there is no callback.
         *
         * The return object should look something like this:
         * {
         *     url: 'http://bidserver.com/api/bids' // base request url for a GET/POST request
         *     data: { // query string object that will be attached to the base url
         *        slots: [
         *             {
         *                 placementId: 54321,
         *                 sizes: [[300, 250]]
         *             },{
         *                 placementId: 12345,
         *                 sizes: [[300, 600]]
         *             },{
         *                 placementId: 654321,
         *                 sizes: [[728, 90]]
         *             }
         *         ],
         *         site: 'http://google.com'
         *     },
         *     callbackId: '_23sd2ij4i1' //unique id used for pairing requests and responses
         * }
         */

        /* ---------------------- PUT CODE HERE ------------------------------------ */
        var placements = [];
        var len = returnParcels.length;
        var identityData;
        var tmp;
        for (var i = 0; i < len; i++) {
            tmp = {
                placementId: returnParcels[i].xSlotRef.placementId,
                sizes: returnParcels[i].xSlotRef.sizes || [],
                traffic: returnParcels[i].xSlotRef.mediaType || 'banner',
                bidId: returnParcels[i].requestId,
                eids: []
            };
            identityData = returnParcels[i] && returnParcels[i].identityData;
            if (identityData) {
                if (identityData.AdserverOrgIp && identityData.AdserverOrgIp.data) {
                    tmp.eids.push(identityData.AdserverOrgIp.data);
                }

                if (identityData.LiveRampIp && identityData.LiveRampIp.data) {
                    tmp.eids.push(identityData.LiveRampIp.data);
                }
            }
            placements.push(tmp);
        }
        var secure = Browser.getProtocol() === 'https:' ? 1 : 0;
        var queryObj = {
            host: Browser.getHostname(),
            page: Browser.getPathname(),
            deviceHeight: Browser.getScreenHeight(),
            deviceWidth: Browser.getScreenWidth(),
            placements: placements,
            secure: secure,
            wrapper: 'index'
        };
        var callbackId = System.generateUniqueId();

        /* Change this to your bidder endpoint. */
        var baseUrl = Browser.getProtocol() + '//colossusssp.com/?c=o&m=multi';

        /* ------------------------ Get consent information -------------------------
         * If you want to implement GDPR consent in your adapter, use the function
         * ComplianceService.gdpr.getConsent() which will return an object.
         *
         * Here is what the values in that object mean:
         *      - applies: the boolean value indicating if the request is subject to
         *      GDPR regulations
         *      - consentString: the consent string developed by GDPR Consent Working
         *      Group under the auspices of IAB Europe
         *
         * The return object should look something like this:
         * {
         *      applies: true,
         *      consentString: "BOQ7WlgOQ7WlgABABwAAABJOACgACAAQABA"
         * }
         *
         * You can also determine whether or not the publisher has enabled privacy
         * features in their wrapper by querying ComplianceService.isPrivacyEnabled().
         *
         * This function will return a boolean, which indicates whether the wrapper's
         * privacy features are on (true) or off (false). If they are off, the values
         * returned from gdpr.getConsent() are safe defaults and no attempt has been
         * made by the wrapper to contact a Consent Management Platform.
         */
        if (ComplianceService) {
            queryObj.gdprStatus = ComplianceService.gdpr.getConsent();
            queryObj.privacyEnabled = ComplianceService.isPrivacyEnabled();
            queryObj.ccpa = ComplianceService.usp && ComplianceService.usp.getConsent();
        }

        /* ---------------- Craft bid request using the above returnParcels --------- */

        /* ------- Put GDPR consent code here if you are implementing GDPR ---------- */

        /* -------------------------------------------------------------------------- */

        return {
            url: baseUrl,
            data: queryObj,
            callbackId: callbackId,
            networkParamOverrides: {
                method: 'POST',
                contentType: 'text/plain'
            }
        };
    }

    /* =============================================================================
     * STEP 3  | Response callback
     * -----------------------------------------------------------------------------
     *
     * This generator is only necessary if the partner's endpoint has the ability
     * to return an arbitrary ID that is sent to it. It should retrieve that ID from
     * the response and save the response to adResponseStore keyed by that ID.
     *
     * If the endpoint does not have an appropriate field for this, set the profile's
     * callback type to CallbackTypes.CALLBACK_NAME and omit this function.
     */
    function adResponseCallback(adResponse) {
        /* Get callbackId from adResponse here */
        var callbackId = 0;
        __baseClass._adResponseStore[callbackId] = adResponse;
    }

    /* -------------------------------------------------------------------------- */

    /* Helpers
     * ---------------------------------- */

    /* =============================================================================
     * STEP 5  | Rendering Pixel
     * -----------------------------------------------------------------------------
     *
    */

    /**
     * This function will render the pixel given.
     * @param  {string} pixelUrl Tracking pixel img url.
     */
    function __renderPixel(pixelUrl) {
        if (pixelUrl) {
            Network.img({
                url: decodeURIComponent(pixelUrl),
                method: 'GET'
            });
        }
    }

    /**
     * Parses and extracts demand from adResponse according to the adapter and then attaches it
     * to the corresponding bid's returnParcel in the correct format using targeting keys.
     *
     * @param {string} sessionId The sessionId, used for stats and other events.
     *
     * @param {any} adResponse This is the bid response as returned from the bid request, that was either
     * passed to a JSONP callback or simply sent back via AJAX.
     *
     * @param {object[]} returnParcels The array of original parcels, SAME array that was passed to
     * generateRequestObj to signal which slots need demand. In this funciton, the demand needs to be
     * attached to each one of the objects for which the demand was originally requested for.
     */
    function __parseResponse(sessionId, adResponse, returnParcels) {
        /* =============================================================================
         * STEP 4  | Parse & store demand response
         * -----------------------------------------------------------------------------
         *
         * Fill the below variables with information about the bid from the partner, using
         * the adResponse variable that contains your module adResponse.
         */

        /* This an array of all the bids in your response that will be iterated over below. Each of
         * these will be mapped back to a returnParcel object using some criteria explained below.
         * The following variables will also be parsed and attached to that returnParcel object as
         * returned demand.
         *
         * Use the adResponse variable to extract your bid information and insert it into the
         * bids array. Each element in the bids array should represent a single bid and should
         * match up to a single element from the returnParcel array.
         *
         */

        /* ---------- Process adResponse and extract the bids into the bids array ------------ */

        var bids = adResponse;

        /* --------------------------------------------------------------------------------- */

        for (var j = 0; j < returnParcels.length; j++) {
            var curReturnParcel = returnParcels[j];

            var headerStatsInfo = {};
            var htSlotId = curReturnParcel.htSlot.getId();
            headerStatsInfo[htSlotId] = {};
            headerStatsInfo[htSlotId][curReturnParcel.requestId] = [curReturnParcel.xSlotName];

            var curBid;

            for (var i = 0; i < bids.length; i++) {
                /**
                 * This section maps internal returnParcels and demand returned from the bid request.
                 * In order to match them correctly, they must be matched via some criteria. This
                 * is usually some sort of placements or inventory codes. Please replace the someCriteria
                 * key to a key that represents the placement in the configuration and in the bid responses.
                 */

                /* ----------- Fill this out to find a matching bid for the current parcel ------------- */
                if (curReturnParcel.xSlotRef.requestId === bids[i].bidId) {
                    curBid = bids[i];
                    bids.splice(i, 1);

                    break;
                }
            }

            /* No matching bid found so its a pass */
            if (!curBid) {
                if (__profile.enabledAnalytics.requestTime) {
                    __baseClass._emitStatsEvent(sessionId, 'hs_slot_pass', headerStatsInfo);
                }
                curReturnParcel.pass = true;

                continue;
            }

            /* ---------- Fill the bid variables with data from the bid response here. ------------ */

            /* Using the above variable, curBid, extract various information about the bid and assign it to
             * these local variables */

            /* The bid price for the given slot */
            var bidPrice = curBid.cpm;

            /* The size of the given slot */
            var bidSize = [Number(curBid.width), Number(curBid.height)];

            /* The creative/adm for the given slot that will be rendered if is the winner.
             * Please make sure the URL is decoded and ready to be document.written.
             */
            var bidCreative = curBid.ad;

            /* The dealId if applicable for this slot. */
            var bidDealId = curBid.dealId;

            /* Explicitly pass */
            var bidIsPass = bidPrice <= 0;

            /* OPTIONAL: tracking pixel url to be fired AFTER rendering a winning creative.
            * If firing a tracking pixel is not required or the pixel url is part of the adm,
            * leave empty;
            */
            var pixelUrl = '';

            /* --------------------------------------------------------------------------------------- */

            curBid = null;
            if (bidIsPass) {
                //? if (DEBUG) {
                Scribe.info(__profile.partnerId + ' returned pass for { id: ' + adResponse.id + ' }.');
                //? }
                if (__profile.enabledAnalytics.requestTime) {
                    __baseClass._emitStatsEvent(sessionId, 'hs_slot_pass', headerStatsInfo);
                }
                curReturnParcel.pass = true;

                continue;
            }

            if (__profile.enabledAnalytics.requestTime) {
                __baseClass._emitStatsEvent(sessionId, 'hs_slot_bid', headerStatsInfo);
            }

            curReturnParcel.size = bidSize;
            curReturnParcel.targetingType = 'slot';
            curReturnParcel.targeting = {};

            var targetingCpm = '';

            //? if (FEATURES.GPT_LINE_ITEMS) {
            targetingCpm = __baseClass._bidTransformers.targeting.apply(bidPrice);
            var sizeKey = Size.arrayToString(curReturnParcel.size);

            if (bidDealId) {
                curReturnParcel.targeting[__baseClass._configs.targetingKeys.pmid] = [sizeKey + '_' + bidDealId];
                curReturnParcel.targeting[__baseClass._configs.targetingKeys.pm] = [sizeKey + '_' + targetingCpm];
            } else {
                curReturnParcel.targeting[__baseClass._configs.targetingKeys.om] = [sizeKey + '_' + targetingCpm];
            }
            curReturnParcel.targeting[__baseClass._configs.targetingKeys.id] = [curReturnParcel.requestId];
            //? }

            //? if (FEATURES.RETURN_CREATIVE) {
            curReturnParcel.adm = bidCreative;
            if (pixelUrl) {
                curReturnParcel.winNotice = __renderPixel.bind(null, pixelUrl);
            }
            //? }

            //? if (FEATURES.RETURN_PRICE) {
            curReturnParcel.price = Number(__baseClass._bidTransformers.price.apply(bidPrice));
            //? }

            var expiry = 0;
            if (__profile.features.demandExpiry.enabled) {
                expiry = __profile.features.demandExpiry.value + System.now();
            }

            var pubKitAdId = RenderService.registerAd({
                sessionId: sessionId,
                partnerId: __profile.partnerId,
                adm: bidCreative,
                requestId: curReturnParcel.requestId,
                size: curReturnParcel.size,
                price: targetingCpm,
                dealId: bidDealId || null,
                timeOfExpiry: expiry,
                auxFn: __renderPixel,
                auxArgs: [pixelUrl]
            });
            //? if (FEATURES.INTERNAL_RENDER) {
            curReturnParcel.targeting.pubKitAdId = pubKitAdId;
            //? }
        }
    }

    /* =====================================
     * Constructors
     * ---------------------------------- */

    (function __constructor() {
        ComplianceService = SpaceCamp.services.ComplianceService;
        RenderService = SpaceCamp.services.RenderService;

        /* =============================================================================
         * STEP 1  | Partner Configuration
         * -----------------------------------------------------------------------------
         *
         * Please fill out the below partner profile according to the steps in the README doc.
         */

        /* ---------- Please fill out this partner profile according to your module ------------ */
        __profile = {
            partnerId: 'ColossusHtb',
            namespace: 'ColossusHtb',
            statsId: 'CLSS',
            version: '2.1.0',
            targetingType: 'slot',
            enabledAnalytics: {
                requestTime: true
            },
            features: {
                demandExpiry: {
                    enabled: false,
                    value: 0
                },
                rateLimiting: {
                    enabled: false,
                    value: 0
                }
            },

            /* Targeting keys for demand, should follow format ix_{statsId}_id */
            targetingKeys: {
                id: 'ix_clss_id',
                om: 'ix_clss_cpm',
                pm: 'ix_clss_cpm',
                pmid: 'ix_clss_dealid'
            },

            /* The bid price unit (in cents) the endpoint returns, please refer to the readme for details */
            bidUnitInCents: 1,
            lineItemType: Constants.LineItemTypes.ID_AND_SIZE,
            callbackType: Partner.CallbackTypes.NONE,
            architecture: Partner.Architectures.SRA,
            requestType: Partner.RequestTypes.ANY
        };

        /* --------------------------------------------------------------------------------------- */

        //? if (DEBUG) {
        var results = ConfigValidators.partnerBaseConfig(configs) || PartnerSpecificValidator(configs);

        if (results) {
            throw Whoopsie('INVALID_CONFIG', results);
        }
        //? }

        __baseClass = Partner(__profile, configs, null, {
            parseResponse: __parseResponse,
            generateRequestObj: __generateRequestObj,
            adResponseCallback: adResponseCallback
        });
    })();

    /* =====================================
     * Public Interface
     * ---------------------------------- */

    var derivedClass = {
        /* Class Information
         * ---------------------------------- */

        //? if (DEBUG) {
        __type__: 'ColossusHtb',
        //? }

        //? if (TEST) {
        __baseClass: __baseClass,
        //? }

        /* Data
         * ---------------------------------- */

        //? if (TEST) {
        profile: __profile,
        //? }

        /* Functions
         * ---------------------------------- */

        //? if (TEST) {
        parseResponse: __parseResponse,
        generateRequestObj: __generateRequestObj
        //? }
    };

    return Classify.derive(__baseClass, derivedClass);
}

////////////////////////////////////////////////////////////////////////////////
// Exports /////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////

module.exports = ColossusHtb;
